server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -315,4 +377,119 @@ protected boolean isFresh(Long currentKey, Long newKey) {
+        }
+
+        @Override
+        protected void onItemResponse(String clusterAlias, RemoteClusterStatsResponse response) {
+            if (isCancelled() == false) {
+                responses.put(clusterAlias, response);
+            }
+        }
+
+        @Override
+        protected void onItemFailure(String clusterAlias, Exception e) {
+            logger.warn("Failed to get remote cluster stats for [{}]: {}", clusterAlias, e);
+        }
+
+        private boolean isCancelled() {
+            return task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled();
+        }
+
+        @Override
+        protected Map<String, RemoteClusterStats> onCompletion() {
+            if (isCancelled()) {
+                return Map.of();
+            }
+
+            Map<String, RemoteClusterStats> remoteClustersStats = new HashMap<>();
+
+            for (String clusterAlias : remoteClusterService.getRegisteredRemoteClusterNames()) {
+                RemoteClusterConnection remoteConnection = remoteClusterService.getRemoteClusterConnection(clusterAlias);
+                RemoteConnectionInfo remoteConnectionInfo = remoteConnection.getConnectionInfo();
+                RemoteClusterStatsResponse response = responses.get(clusterAlias);
+                var compression = RemoteClusterService.REMOTE_CLUSTER_COMPRESS.getConcreteSettingForNamespace(clusterAlias).get(settings);
+                var remoteClusterStats = new RemoteClusterStats(
+                    response,
+                    remoteConnectionInfo.getModeInfo().modeName(),
+                    remoteConnection.isSkipUnavailable(),
+                    compression.toString()
+                );
+                remoteClustersStats.put(clusterAlias, remoteClusterStats);
+            }
+            return remoteClustersStats;
+        }
+    }
+
+    private class ActionListenerWithRemotes implements ActionListener<ClusterStatsResponse> {
+        private final ActionListener<ClusterStatsResponse> listener;
+        private final SubscribableListener<Map<String, RemoteClusterStats>> remoteListener;
+
+        ActionListenerWithRemotes(Task task, ClusterStatsRequest request, ActionListener<ClusterStatsResponse> listener) {
+            this.listener = listener;
+            remoteListener = getStatsFromRemotes(task, request);
+        }
+
+        SubscribableListener<Map<String, RemoteClusterStats>> getStatsFromRemotes(Task task, ClusterStatsRequest request) {
+            if (doRemotes(request) == false) {
+                return SubscribableListener.newSucceeded(null);
+            }
+            var remotes = remoteClusterService.getRegisteredRemoteClusterNames();
+            if (remotes.isEmpty()) {
+                return SubscribableListener.newSucceeded(Map.of());
+            }
+            var remotesFuture = new ListenableFuture<Map<String, RemoteClusterStats>>();
```

Comment:
Do you need the exception-mangling behaviour of `ListenableFuture` here? Do we need to return a future at all? Normally async methods accept an `ActionListener<>` parameter to avoid constructing an intermediate future.

Reply:
We actually only need `SubscribableListener` there, the future type remains from older code, I'll change it.